File: backend/app/schemas/agents.py
from datetime import datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.body import BodyRegion
from app.schemas.intake import Aggravator, Reliever
from app.schemas.red_flags import RedFlagHit

# =====================================================================
# Agent 1 — Extractor
# =====================================================================


class ReportedDiagnosisEvidence(BaseModel):
    """Diagnostic term extracted VERBATIM from a document. Evidence of what
    someone wrote, NOT something this system diagnoses.
    """

    term: str
    source_file_id: str = Field(alias="sourceFileId")
    verbatim_quote: str = Field(alias="verbatimQuote")

    model_config = {"populate_by_name": True}


class ExtractionProvenance(BaseModel):
    file_id: str = Field(alias="fileId")
    extraction_model: str = Field(alias="extractionModel")
    confidence: float

    model_config = {"populate_by_name": True}


class ExtractedClinicalPayload(BaseModel):
    primary_location: BodyRegion = Field(alias="primaryLocation")
    secondary_locations: list[BodyRegion] = Field(default_factory=list, alias="secondaryLocations")
    aggravators: list[Aggravator] = Field(default_factory=list)
    relievers: list[Reliever] = Field(default_factory=list)
    measurements: dict[str, float | str] = Field(default_factory=dict)
    reported_diagnoses: list[ReportedDiagnosisEvidence] = Field(
        default_factory=list, alias="reportedDiagnoses"
    )
    provenance: list[ExtractionProvenance] = Field(default_factory=list)
    unclassified_notes: list[str] = Field(default_factory=list, alias="unclassifiedNotes")

    model_config = {"populate_by_name": True}


# =====================================================================
# Agent 2 — Triage
# =====================================================================


class TriageDisposition(StrEnum):
    proceed = "proceed"
    halt_seek_emergency = "halt_seek_emergency"
    halt_seek_physician = "halt_seek_physician"
    proceed_with_caution = "proceed_with_caution"


class TriageResult(BaseModel):
    disposition: TriageDisposition
    hits: list[RedFlagHit] = Field(default_factory=list)
    rationale: str
    evaluated_at: datetime = Field(alias="evaluatedAt")
    evaluated_by: str = Field(alias="evaluatedBy")

    model_config = {"populate_by_name": True}

    @property
    def halted(self) -> bool:
        return self.disposition in (
            TriageDisposition.halt_seek_emergency,
            TriageDisposition.halt_seek_physician,
        )


# =====================================================================
# Agent 3 — RAG Retriever + plan composer
# =====================================================================


ExercisePhase = Literal[
    "isometric_stabilization",
    "controlled_range_of_motion",
    "loaded_mobility",
    "integrated_movement",
]


class ExerciseMediaPlaceholder(BaseModel):
    caption: str


class ExerciseCitation(BaseModel):
    source_title: str = Field(alias="sourceTitle")
    chunk_id: str = Field(alias="chunkId")
    score: float

    model_config = {"populate_by_name": True}


class Exercise(BaseModel):
    id: str
    name: str
    phase: ExercisePhase
    target_region: BodyRegion = Field(alias="targetRegion")
    instructions: list[str]
    dose: str
    stop_conditions: list[str] = Field(alias="stopConditions")
    contraindication_tags: list[str] = Field(default_factory=list, alias="contraindicationTags")
    media_placeholders: dict[str, ExerciseMediaPlaceholder] = Field(
        default_factory=dict, alias="mediaPlaceholders"
    )
    citations: list[ExerciseCitation] = Field(default_factory=list)

    model_config = {"populate_by_name": True}


class PlanPhase(BaseModel):
    phase: ExercisePhase
    summary: str
    exercises: list[Exercise]

    model_config = {"populate_by_name": True}


class ProbabilisticFraming(BaseModel):
    """Hard-coded probabilistic framing. The ONLY place the system talks
    about what symptoms might be consistent with — and never as a diagnosis.
    """

    pattern: str
    commonly_associated_with: list[str] = Field(alias="commonlyAssociatedWith")
    confidence: Literal["low", "moderate", "high"]

    model_config = {"populate_by_name": True}


class RehabPlan(BaseModel):
    id: str
    generated_at: datetime = Field(alias="generatedAt")
    phases: list[PlanPhase]
    probabilistic_framing: ProbabilisticFraming = Field(alias="probabilisticFraming")
    clinician_reviewed: bool = Field(default=False, alias="clinicianReviewed")
    clinician_note: str | None = Field(default=None, alias="clinicianNote")

    model_config = {"populate_by_name": True}

File: lib/types/agents.ts
import type { BodyRegion } from "./body";
import type { Aggravator, Reliever } from "./intake";
import type { RedFlagHit } from "./redFlags";

/**
 * Contracts between the orchestrator and each agent. The frontend never
 * produces these — it only consumes them once the backend populates them
 * on the IntakeSubmission record. Defined here so the UI can render
 * results type-safely and the same types can be shared with the Python
 * FastAPI backend via a codegen step later.
 */

// =====================================================================
// Agent 1 — The Extractor
// =====================================================================

/**
 * Normalized clinical payload produced by the Extractor. Merges the
 * user's subjective input with structured evidence from uploaded files
 * (VLM for images, text parsing for reports).
 *
 * Numeric findings go in `measurements` (e.g., LCEA angle, Cobb angle,
 * ROM degrees). Free-text diagnoses that appear in uploaded reports go
 * in `reportedDiagnoses` — these are evidence of what *someone wrote*,
 * NOT a diagnosis this system is making.
 */
export interface ExtractedClinicalPayload {
  primaryLocation: BodyRegion;
  secondaryLocations: BodyRegion[];
  aggravators: Aggravator[];
  relievers: Reliever[];
  /**
   * Structured numeric findings keyed by lowercase snake_case term.
   * Examples: lcea_left, cobb_angle, knee_flexion_degrees.
   */
  measurements: Record<string, number | string>;
  /**
   * Diagnostic terms extracted verbatim from clinical documents.
   * These are EVIDENCE, not conclusions this system draws.
   */
  reportedDiagnoses: Array<{
    term: string;
    sourceFileId: string;
    verbatimQuote: string;
  }>;
  /** Per-file confidence & provenance. */
  provenance: Array<{
    fileId: string;
    extractionModel: string;
    confidence: number; // 0..1
  }>;
  /**
   * Anything the Extractor saw but could not classify. Gets shown to the
   * clinician at HITL review so nothing silently gets dropped.
   */
  unclassifiedNotes: string[];
}

// =====================================================================
// Agent 2 — The Triage
// =====================================================================

export type TriageDisposition =
  | "proceed"
  | "halt_seek_emergency"
  | "halt_seek_physician"
  | "proceed_with_caution";

export interface TriageResult {
  disposition: TriageDisposition;
  hits: RedFlagHit[];
  /** Human-readable rationale — shown to the user if the pipeline halts. */
  rationale: string;
  /** ISO timestamp of the triage decision. */
  evaluatedAt: string;
  /** Model identifier used (e.g., "nemotron-3-8b-safety-2026-03"). */
  evaluatedBy: string;
}

// =====================================================================
// Agent 3 — The RAG Retriever & plan composer
// =====================================================================

export type ExercisePhase =
  | "isometric_stabilization"
  | "controlled_range_of_motion"
  | "loaded_mobility"
  | "integrated_movement";

export interface Exercise {
  id: string;            // stable ID from the knowledge base
  name: string;
  phase: ExercisePhase;
  targetRegion: BodyRegion;
  /** Educational instructions — plain text, phrased as cues not commands. */
  instructions: string[];
  /** Dose guidance, e.g. "2 sets of 10 holds, 10s each". */
  dose: string;
  /** Cues to stop: pain spikes, paresthesia, etc. */
  stopConditions: string[];
  /**
   * Contraindication tags matched against the user's profile at
   * retrieval time. If any of these are active for the user, the
   * exercise is NOT returned.
   */
  contraindicationTags: string[];
  mediaPlaceholders: {
    video?: { caption: string };
    image?: { caption: string };
  };
  /** RAG source document chunks used to justify this pick. */
  citations: Array<{
    sourceTitle: string;
    chunkId: string;
    score: number;
  }>;
}

export interface RehabPlan {
  id: string;
  generatedAt: string;
  /**
   * Ordered phases — the user is never shown a plan "menu"; they move
   * through phases under the logic the clinician cleared.
   */
  phases: Array<{
    phase: ExercisePhase;
    summary: string;
    exercises: Exercise[];
  }>;
  /**
   * Probabilistic framing for the UI. This is the ONLY place the system
   * describes what the user's symptoms might be consistent with — and
   * even then, never as "You have X."
   */
  probabilisticFraming: {
    pattern: string;                 // e.g., "anterior hip impingement pattern"
    commonlyAssociatedWith: string[]; // lay terms, never presented as a dx
    confidence: "low" | "moderate" | "high";
  };
  /** Set by HITL workflow — plan not released to the user until true. */
  clinicianReviewed: boolean;
  clinicianNote?: string;
}
